// webdriverio/build/commands/element/click.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = click;

async function click(options) {
  let {
    button
  } = options || {};

  if (button === 'left') {
    button = 0;
  }

  if (button === 'middle') {
    button = 1;
  }

  if (button === 'right') {
    button = 2;
  }

  if (typeof button === 'undefined') {
    return this.elementClick(this.elementId);
  }

  if ([0, 1, 2].includes(button)) {
    if (this.isW3C) {
      await this.performActions([{
        type: 'pointer',
        id: 'pointer1',
        parameters: {
          pointerType: 'mouse'
        },
        actions: [{
          type: 'pointerMove',
          origin: this,
          x: 0,
          y: 0
        }, {
          type: 'pointerDown',
          button
        }, {
          type: 'pointerUp',
          button
        }]
      }]);
      return this.releaseActions();
    }

    await this.moveTo();
    return this.positionClick(button);
  }

  throw new Error('Button type not supported.');
}
